### PR TITLE
Fix WooPay trial subscriptions purchases

### LINKED ISSUE #9777: Free trial subscriptions can no longer be processed using WooPay
### Describe the bug

(copied from the WooPay repo, text by @malithsen)

When placing a free-trial subscription, WooPay shows the following error
Image

These orders seems to be failing due to an order ID mismatch in
Image

### To Reproduce

1. As a merchant, create a free trial subscription
2. As a logged in shopper, purchase the above subscription using WooPay
3. See error when clicking on the "Place order" button in WooPay.

### Actual behavior

Shoppers are unable to purchase free trial subscriptions using WooPay

### Expected behavior

Shoppers should be able to purchase free trial subscriptions using WooPay.

### PR DESCRIPTION
**Pull request by alefesouza**
Fixes #9777

#### Changes proposed in this Pull Request

This PR fixes an error that happens when purchasing only trial subscriptions with WooPay.

The error happens on [this line](https://github.com/woocommerce/woocommerce/blob/1675618a26dc8d2d1026e14d252b9bef80df30c0/plugins/woocommerce/src/StoreApi/Utilities/DraftOrderTrait.php#L64) because WooPay sets the order to a `processing` state during the preflight request and orders with trial subscription has a `0` total price, making the `need_payment()` return `false` because [it expects the order status](https://github.com/woocommerce/woocommerce/blob/1675618a26dc8d2d1026e14d252b9bef80df30c0/plugins/woocommerce/includes/class-wc-order.php#L1787) to be `pending` or `failed`, thus creating a new order making we get the order id mismatch error [here](https://github.com/Automattic/woocommerce-payments/blob/393190472bb2a5c50f9e7addaac6f45018c63841/includes/class-wc-payment-gateway-wcpay.php#L1747).

#### Testing instructions

* Add one or more trial subscriptions to the cart.
* Checkout with WooPay.
* Place order on WooPay.
* The order should be placed successfully.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

File: includes/class-wc-payment-gateway-wcpay.php [hunk 1]
```diff
@@ -1241,6 +1241,11 @@ public function process_payment( $order_id ) {
 				// Set the order status to "pending payment".
 				$order->update_status( 'pending' );
 
+				// phpcs:ignore WordPress.Security.NonceVerification.Missing
+				if ( isset( $_POST['order_has_subscriptions'] ) ) {
+					$order->update_meta_data( 'order_has_subscriptions', '1' );
+				}
```
Comment (alefesouza) [inline includes/class-wc-payment-gateway-wcpay.php]: During the preflight check the draft order has no subscription metadata, this is the only approach I found to identify later if the order is a subscription.

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9778` or branch name `fix/woopay-trial-subscriptions` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/woopay-trial-subscriptions) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 27990a4899f5df84a9c5d50b1809edc6e55efda0
- Build time: 2024-11-23 03:18:26 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/class-wc-payment-gateway-wcpay.php (hunk 1 above)
Comment (alefesouza) [reply includes/class-wc-payment-gateway-wcpay.php]: I was able to remove it on https://github.com/Automattic/woocommerce-payments/pull/9778/commits/27990a4899f5df84a9c5d50b1809edc6e55efda0 .

**Review by leonardola — APPROVED** [review]
Looks good to me.

Tested with trial subscriptions, regular subscriptions, regular items, mixed items.

**Pull request merged**